### PR TITLE
 client/network-gossip/src/bridge: Finish when network event stream closes

### PR DESCRIPTION
**Pull request by mxinden**
Previously within `<GossipEngine as Future>::poll` one would poll the
`network_event_stream` ignoring all messages other than
`Poll::Ready(Some())`. Ignoring `Poll::Ready(None)` leads to a panic on
the next poll of the stream, gien that it is not fused.

By design `network_event_stream` does not close unless an unbounded send
into it fails, or the `NetworkWorker` gets shut down.

> The stream never ends (unless the `NetworkWorker` gets shut down).
> (client/network/src/service.rs)

An `unbounded_send` to fail on an unbounded channel is unlikely. The
`NetworkWorker` shutting down is not unlikely. In such case the
`GossipEngine` should shut down as well.

With this patch a `<GossipEngine as Future>` finishes on
`Poll::Ready(None)` returned from `network_event_stream`.

In addition this pull request contains a second commit:

Have `<NetworkBridge as Future>::poll` return `Poll::Ready(Err)` instead
of `Poll::Ready(Ok)` to be consistent with the handling of the neighbor
packet worker stream and the gossip validator report stream. Both `Err`
as well as `Ok` shut down the `NetworkBridge` as well as the
`VoterWorker`.

Relates to https://github.com/paritytech/substrate/issues/5000.

### CODE REVIEW

**Comment by bkchr** [comment]
Shouldn't this fixed the linked issue?

Any idea how we can test this?

**Comment by mxinden** [comment]
> Shouldn't this fixed the linked issue?

Very likely, yes. I didn't write `Fixes xxx` to still keep an eye on it. We can close the issue if you like @bkchr.

**Comment by bkchr** [comment]
> Any idea how we can test this?
? :D

**Comment by mxinden** [comment]
Regression test added via https://github.com/paritytech/substrate/pull/5282/commits/a422fba42f150b375d94149e52561f352685acea.

@bkchr thanks for following up on this.

File: client/network-gossip/src/bridge.rs [hunk 1]
```diff
@@ -39,6 +39,7 @@ impl<B: BlockT> Unpin for GossipEngine<B> {}
 
 impl<B: BlockT> GossipEngine<B> {
 	/// Create a new instance.
+	// TODO: We don't need this clone, right?
```
Comment (rphmeier) [inline client/network-gossip/src/bridge.rs]: stray TODO

File: client/network-gossip/src/bridge.rs [hunk 2]
```diff
@@ -148,8 +149,8 @@ impl<B: BlockT> Future for GossipEngine<B> {
 	fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
 		let this = &mut *self;
 
-		while let Poll::Ready(Some(event)) = this.network_event_stream.poll_next_unpin(cx) {
-			match event {
+		loop { match this.network_event_stream.poll_next_unpin(cx) {
```
Comment (rphmeier) [inline client/network-gossip/src/bridge.rs]: nit: loop body should gain 1 level indentation. I find that more readable

File: client/network-gossip/Cargo.toml [hunk 3]
```diff
@@ -18,4 +18,5 @@ log = "0.4.8"
 lru = "0.4.3"
 sc-network = { version = "0.8.0-alpha.2", path = "../network" }
 sp-runtime = { version = "2.0.0-alpha.2", path = "../../primitives/runtime" }
+substrate-test-runtime-client = { version = "2.0.0-dev", path = "../../test-utils/runtime/client" }
```
Comment (bkchr) [inline client/network-gossip/Cargo.toml]: This needs to go to `dev-dependencies`!

**Comment by bkchr** [comment]
This revealed another flaky grandpa test:
```
failures:
---- observer::tests::observer_work_polls_underlying_network_bridge stdout ----
thread 'observer::tests::observer_work_polls_underlying_network_bridge' panicked at 'assertion failed: observer.now_or_never().is_none()', client/finality-grandpa/src/observer.rs:441:13
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1052
   5: std::io::Write::write_fmt
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447/src/libstd/io/mod.rs:1426
   6: std::io::impls::<impl std::io::Write for alloc::boxed::Box<W>>::write_fmt
             at src/libstd/io/impls.rs:156
   7: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   8: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   9: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
  10: std::panicking::default_hook
             at src/libstd/panicking.rs:221
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:472
  12: std::panicking::begin_panic
  13: <std::future::GenFuture<T> as core::future::future::Future>::poll
  14: std::thread::local::LocalKey<T>::with
  15: futures_executor::local_pool::block_on
  16: sc_finality_grandpa::observer::tests::observer_work_polls_underlying_network_bridge
  17: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447/src/liballoc/boxed.rs:1015
  18: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:86
  19: std::panicking::try
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447/src/libstd/panicking.rs:281
  20: std::panic::catch_unwind
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447/src/libstd/panic.rs:394
  21: test::run_test_in_process
             at src/libtest/lib.rs:539
  22: test::run_test::run_test_inner::{{closure}}
             at src/libtest/lib.rs:452
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
failures:
    observer::tests::observer_work_polls_underlying_network_bridge
```

**Comment by mxinden** [comment]
@bkchr @rphmeier could you take another look?

**Pull request merged**